GDScript: Implement pattern guards for match statement

vnen opened the pull request:
Within a match statement, it is now possible to add guards in each branch:

	var a = 0
	match a:
		0 when false: print("does not run")
		0 when true: print("but this does")

This allows more complex logic for deciding which branch to take.

Closes https://github.com/godotengine/godot-proposals/issues/4775

**EDIT:** Changed the `-> if` separator to the new keyword `when`, as discussed.

vnen committed:
GDScript: Implement pattern guards for match statement

Within a match statement, it is now possible to add guards in each
branch:

	var a = 0
	match a:
		0 when false: print("does not run")
		0 when true: print("but this does")

This allows more complex logic for deciding which branch to take.

dalexeev commented on modules/gdscript/gdscript_parser.cpp:
```diff
@@ -1997,7 +1997,37 @@ GDScriptParser::MatchBranchNode *GDScriptParser::parse_match_branch() {
+		// Pattern guard.
+		// Create block for guard because it also needs to access the bound variables from patterns, and we don't want to add them to the outer scope.
+		branch->guard_body = alloc_node<SuiteNode>();
+		for (const KeyValue<StringName, IdentifierNode *> &E : branch->patterns[0]->binds) {
```
This line causes the editor crash if the pattern is not completed.

```gdscript
match x:
    var -> if y == 0:
```

(After applying the patch on the current master, line 2005 became line 2018.)

```
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.2.dev.custom_build (3e153f026ab9734df61631a96b412d40008e7919)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x3c4b0) [0x7f949c03c4b0] (??:0)
[2] CowData<GDScriptParser::PatternNode*>::get(int) const (/home/danil/godot-src/./core/templates/cowdata.h:155 (discriminator 7))
[3] Vector<GDScriptParser::PatternNode*>::operator[](int) const (/home/danil/godot-src/./core/templates/vector.h:96)
[4] GDScriptParser::parse_match_branch() (/home/danil/godot-src/modules/gdscript/gdscript_parser.cpp:2018)
[5] GDScriptParser::parse_match() (/home/danil/godot-src/modules/gdscript/gdscript_parser.cpp:1957)
```

vnen replied:
That's a bad copy-paste of mine, nice catch

dalexeev reviewed (commented):
The code looks good to me! :+1: There is a little hack with the pattern guard is `SuiteNode`, not `ExpressionNode`. In theory, we can get unrelated warnings or something like that, but I did not find any related bugs when I tested it.

The tests are good, but maybe we should add some more? For example a test that an assignment is not allowed inside a pattern guard.

I'm still not sure about the syntax, so I suggest discussing this at the next GDScript team meeting.

vnen commented:
> I noticed that the pattern guard is executed even if none of the patterns matched.
> 
> ```gdscript
> func f() -> bool:
>     print("f")
>     return true
> 
> func _ready() -> void:
>     var x := 1
>     match x:
>         2 -> if f():
>             print("A")
>         _:
>             print("B")
> ```
> 
> Prints:
> 
> ```
> f
> B
> ```

Fixed now.	Added a test for this case as well as the one for assignment.

adamscott commented:
Discussed during a GDScript Team Meeting: this PR will be modified my @vnen to use the `when` keyword to introduce the if statement. This PR would be ready for 4.2 afterwards.

dalexeev reviewed (approved):
Looks good to me and works great, it won't even break compatibility if someone uses `when` as an identifier. Thank you George!

```gdscript
var when = 1
match when:
    when when when: # I like it.
        when
```

YuriSizov commented:
Thanks!
